core/src/main/scala/org/apache/spark/scheduler/dynalloc/ExecutorMonitor.scala

```diff
@@ -137,6 +254,21 @@ private[spark] class ExecutorMonitor(
     val executorId = event.taskInfo.executorId
     val exec = executors.get(executorId)
     if (exec != null) {
+      // If the task succeeded and the stage generates shuffle data, record that this executor
+      // holds data for the shuffle. Note that this ignores speculation, since this code is not
+      // directly tied to the map output tracker that knows exactly which shuffle blocks are
```

Comment:
I think this comment about speculation is a bit confusing.  It sounds a bit like you're saying that speculative tasks are ignored. Really, you are saying is that if there is a speculative and non-speculative task, and both succeed, you say both have shuffle data.  (right?)

and I actually think both tasks will write shuffle data locally and update that state in the driver -- its just that scheduler only sends one location when it sends out the next downstream tasks.  But if the stage fails, it might resubmit immediately using the other copy (but that is off the top of my head, I need to step through that part carefully ...)

Reply:
> if there is a speculative and non-speculative task, and both succeed, you say both have shuffle data

 Right.

> scheduler only sends one location when it sends out the next downstream tasks

IIRC the map output tracker only keeps one location for each shuffle block.

```
  def addMapOutput(mapId: Int, status: MapStatus): Unit = synchronized {
    if (mapStatuses(mapId) == null) {
      _numAvailableOutputs += 1
      invalidateSerializedMapOutputStatusCache()
    }
    mapStatuses(mapId) = status
  }
```

BTW that code is interesting in that it seems like it's possible to have the tracker pointing at one executor while the serialized version is pointing at another... so I guess, inadvertently, the behavior the comment is talking about is "the right thing".

Reply:
ah you are right.  I guess I was remembering pre-SPARK-20715.